### ISSUE TITLE
doc build dependency doesn't work in Visual Studio

### ISSUE THREAD

**Issue by bvanevery** [urho3d/Urho3D#908]
In Visual Studio 2015 Community Edition, when I build the "doc" target before building anything else in the solution, it fails because build\Urho3D\bin\tool\ScriptCompiler doesn't exist yet.  A dependency on ScriptCompiler should exist.  weitjong cannot reproduce the problem in a Makefile based build on a non-Windows system (probably Linux).  The expected dependencies are:
  doc -> ScriptAPI.dox -> Urho3D & ScriptCompiler
and weitjong says it works on his system.  I haven't checked a non-Windows system of my own yet.

**Comment by cadaver** [comment]
It looks like the dependencies (Urho3D, ScriptCompiler, tolua++) should be added with add_dependencies command to the doc target (http://cmake.org/cmake/help/v3.0/command/add_dependencies.html).

**Comment by weitjong** [comment]
@bvanevery You can try to fix the dependency yourself as suggested by cadaver. If it works then you can let us know or send us a pull request.

@cadaver I do not know for sure yet why it does not work for VS 2015. The dependency setup is already there or otherwise we would have this problem on Linux as well.

The ```doc``` target depends on the generated ```ScriptAPI.dox``` [here](https://github.com/urho3d/Urho3D/blob/master/Docs/CMakeLists.txt#L109-L113).
The yet-to-be-generated ```ScriptAPI.dox``` file then depends on ```ScriptCompiler``` [here](https://github.com/urho3d/Urho3D/blob/master/Docs/CMakeLists.txt#L75-L79). In fact it depends on Urho3D library too for obvious reason.

I could not see any reason why this won't work on VS 2015 or VS in general. Having said that there is no harm to try to set the dependency explicitly using add_dependencies() command.

**Comment by bvanevery** [comment]
I remember some kind of historical funkiness of different target types in CMake, as far as specifying dependencies goes.  You had to do a different drill for different cases, it wasn't a uniform system.  But it has been too long for me to remember exactly.  The behavior of add_dependencies() vs. a DEPENDS clause in ADD_CUSTOM_COMMAND is probably rather different though.  A Makefile output might mask the differences for some reason.  Maybe the funkiness is a notion actually needed by Visual Studio for some reason, but not a Makefile.

**Comment by weitjong** [comment]
I have to disagree. If the DEPENDS does not work as I think it should work then we will have a lot more problems on our VS build now because we have hardly used any add_dependencies() explicitly in all our CMakeLists.txt. We do have some if-endif switches for different IDE/compiler for numerous reasons, but setting up dependency is not one of them. Not yet anyway.

**Comment by bvanevery** [comment]
Well I'm not personally ready to re-learn these CMake nuances, I need to get something done of a 3d nature with Urho first, rather than worry about building the docs.  Take my feedback on where the problem lies for what it's worth: a guess.  I just know that historically, nuances did exist.  I'll revisit the gore when it's more appropriate for me to.

**Comment by cadaver** [comment]
@weitjong You're right, the lack of add_dependencies() doesn't seem to be the problem. In an in-source build, I see the project dependencies populated correctly. Will check a clean out-of-source build next.

**Comment by bvanevery** [comment]
When you say "correctly set", do you mean that starting with a clean slate, you actually fired up the .sln for all of Urho3d, selected "doc" project only, built only that, and everything just worked?  Bear in mind also that the build currently pollutes the source tree per issue #879.

Which VS 2015?  I'm using Community Edition.
My CMake is 3.2.3.

**Comment by cadaver** [comment]
I meant that I observed the checkboxes for Urho3D & ScriptCompiler dependencies being ticked for "doc" project. And yes, a build of the "doc" project only would succeed.

I'm also using VS Community.

However, I'm able to reproduce "doc" build failing due to missing ScriptCompiler, if the build is 64-bit, and the ScriptAPI.dox & LuaScriptAPI.dox files have been deleted from the source tree beforehand. I'm also seeing an unexpected return value from ScriptCompiler (?) on a 64-bit build, so something is messed up.

**Comment by bvanevery** [comment]
Well that's interesting.  My CMake output generator is "Visual Studio 14 2015 Win64".

**Comment by bvanevery** [comment]
I'm seeing some odd logic in Urho3D/Docs/CMakeLists.txt

    if (URHO3D_DOCS)
        set (ALL ALL)
    endif ()

Why is ALL set to ALL?  That does nothing.  Looks like a typo.

**Comment by cadaver** [comment]
The odd return code on 64bit build was something extremely stupid: I had copied the 32bit version of d3dcompiler_47.dll to Windows directory, which caused the FindDirect3D module to always use it and fail to execute a 64bit ScriptCompiler.

**Comment by bvanevery** [comment]
You mean a Windows system directory, definitely not owned by you the user?  Pretty sure that's a big no-no.  Seem to remember reading an article recently about archaic attempts to solve missing DirectX .dll problems by people installing that old hoary DirectX end user runtime.  It's not something people are supposed to do and maybe on later versions of Windows, like 10, it could result in problems?  Not sure.  I know I recently ran into a missing d3dcompiler_47.dll problem recently on one of my Windows 10 systems and I can't remember what, if anything, I did about it.  I certainly don't remember what the best practice was supposed to be.  Maybe I had a bad Windows 10 SDK install or something and fixed it that way.

**Comment by bvanevery** [comment]
I suppose set(ALL ALL) sets the variable ALL to the literal string "ALL".  However I'm not seeing anywhere that ALL is initialized, so I wonder if that can trigger undefined consequences.  Like, say, screwing up the dependency of the subsequent ADD_CUSTOM_TARGET(doc ${ALL} ...).  I can't remember if ALL is a recognized CMake global variable that can be exploited for side effects.  It surely seems like a bad practice in any event.

**Comment by cadaver** [comment]
After fixing the d3dcompiler problem (making sure that a 64bit ScriptCompiler can execute) I'm no longer able to reproduce the problem, even on a 64bit clean build with the .dox files deleted beforehand.

**Comment by bvanevery** [comment]
Ok well when you have that posted somewhere I'll try it myself.

**Comment by cadaver** [comment]
The fix for my situation was del c:\windows\d3dcompiler_47.dll :)

In case you're actually experiencing the same issue, which is 64bit ScriptCompiler failing with a dialog "this application cannot start.." you can try switching to OpenGL build, or checking the value of DIRECT3D_DLL and verifying it's coming from a x64 subdirectory such as C:/Program Files (x86)/Windows Kits/8.1/Redist/D3D/x64/d3dcompiler_47.dll

**Comment by bvanevery** [comment]
On Windows 10 I have a C:\Windows\SysWOW64\D3DCompiler_47.dll .  In the CMake build, DIRECT3D_DLL=C:/Windows/System32/d3dcompiler_47.dll .  I find that C:\Windows\System32\D3DCompiler_47.dll does exist.  I have no idea whether it should, on a clean system, so I'm not going to delete it, unless I can find a credible source that shows it's wrong to have it.  I think in the wild though this proves that users *will* have it.  So what is correct build behavior?

It looks like Urho3D\CMake\Modules\FindDirect3D.cmake would need some updating.  For instance I see registry searches for 8.1 and 8.0 SDKs, but my machine actually has v10.0, v8.1A, v8.1, v8.0A, v8.0, and v7.1A.  This machine started life on Windows Vista.  I think it got clean installed to Windows 7.  I know I put VS 2013 Community Edition on it.  Then I dirty upgraded to Windows 10.  Then I installed VS 2015 Community Edition on it.  I might have also tried to throw a standalone Windows 10 SDK on it when something didn't seem to be working during all the upgrading.  I've still got VS 2013 and 2015 working side by side, there's been no reason to remove the older one.  Oh yeah I seem to recall a fakeout sorta VS 2012 as well, it got renamed subsequently to 2013 Community Edition.  Fun fun fun.  Well, blah, looking at that stuff is giving me a headache.  Probably needs a new bug filed.

Will try changing DIRECT3D_DLL and see what happens.

**Comment by bvanevery** [comment]
I remember now!  The article said you're supposed to ship redists like d3dcompiler_47.dll in your own app.  Not stick 'em in system directories or do special installs.  This has been true for a fair number of years it seems.  Read that article something like 3 years ago I think, and it was old news then.

**Comment by cadaver** [comment]
Committed a change which should prevent system path locations being searched by CMake for the compiler DLL. Now it should be fine to have the DLL in C:\Windows or C:\Windows\System32 and it won't be used erroneously.

**Comment by weitjong** [comment]
It is good to know that the root cause of the problem is not the dependency setup itself because if that's true then we would have a lot of work to do to rectify it!

From the last few comments I suppose the root cause has been identified to be faulty DIRECT3D_DLL detection logic which made it possible to find a wrong dll in the system default path and a commit has been pushed to rectify it. I suppose we can consider this as case closed?

> Why is ALL set to ALL? That does nothing. Looks like a typo.

While I am here. I would like to take the time to demystify the ```set (ALL ALL)``` or ```set (JIT JIT)``` technique. No, that's not a typo. It is a technique that we use a number of times in our CMakeLists.txt. The variable is set and only set when a condition is met. When it is not met then the variable is empty. So, in later part of the CMakeLists.txt the construct such as:
```
add_custom_target (doc ${ALL}
```
is actually being read by CMake as either one of these (depends on the value of the variable set):
```
add_custom_target (doc
``` 
or
```
add_custom_target (doc ALL
```
Now that 'ALL' literal would make a big difference to ```add_custom_target()``` command (see http://cmake.org/cmake/help/v3.3/command/add_custom_target.html). 

I am not sure you can find this technique to be used elsewhere. We actually come up with a few CMake techniques ourselves unique to Urho3D project.

**Comment by bvanevery** [comment]
For set(ALL ALL) I figured the intent out, above.  My concern is that CMake may not guarantee that an undeclared variable is empty, or that it's going to continue to be empty between configuration iterations.  I haven't checked whether CMake has any spec for that.

**Comment by bvanevery** [comment]
I don't think the caper is solved by cadaver's commit.  I've got 
DIRECT3D_DLL=C:/Program Files (x86)/Windows Kits/8.1/Redist/D3D/x64/d3dcompiler_47.dll
which is probably a reasonable redistributable to work from.  On a clean build I still get
1>  'C:\Users\bvanevery\devel\build\Urho3D\bin\tool\ScriptCompiler' is not recognized as an internal or external command,
1>  operable program or batch file.
And indeed, it doesn't exist.
CMake log files seem obtuse.

**Comment by cadaver** [comment]
When that error happens, do you have ScriptCompiler ticked as doc's build dependency?

**Comment by bvanevery** [comment]
I have:
doc
  References
    ScriptCompiler

I'm not exactly convinced that's a dependency.  For instance, External Dependencies is empty.  But I'm no Visual Studio expert.

**Comment by bvanevery** [comment]
Tried building only ScriptCompiler.  Gives error:
1>C:\Users\bvanevery\devel\Urho3D\Source\Tools\ScriptCompiler\ScriptCompiler.cpp(23): fatal error C1083: Cannot open include file: 'Urho3D/AngelScript/Script.h': No such file or directory

**Comment by weitjong** [comment]
> For set(ALL ALL) I figured the intent out, above. My concern is that CMake may not guarantee that an undeclared variable is empty, or that it's going to continue to be empty between configuration iterations. I haven't checked whether CMake has any spec for that.

If one day CMake decides to declare that variable name and initialize it for some other usage then yes, it will break our build. But the chance for that to happen is remotely small that I could probably have a better odd winning a lottery than that. :)

**Comment by bvanevery** [comment]
No a namespace collision is not the issue I'm referring to.  The issue is whether undeclared variables are guaranteed to be empty.  I don't know that they are.  The old "you didn't initialize your variables" paranoia.

BTW you would be worrying about some third party using "ALL" as a variable in their own code, not official CMake code.  Like some Urho3d user likes your style, makes a contribution, and people are wondering why things go kablooey.  Or someone borrows your code, but I guess that's not strictly your problem.  Anyways there's a reason people often use somewhat longer variable names.

**Comment by bvanevery** [comment]
I am starting to think that *most* targets in a Visual Studio IDE will not in fact build when you select them "Project Only".  I've been chasing a lot of things backwards and finding they won't build, and give errors about what they're expecting that doesn't exist yet.

**Comment by weitjong** [comment]
> No a namespace collision is not the issue I'm referring to. The issue is whether undeclared variables are guaranteed to be empty. I don't know that they are. The old "you didn't initialize your variables" paranoia.

BTW you would be worrying about some third party using "ALL" as a variable in their own code, not official CMake code. Like some Urho3d user likes your style, makes a contribution, and people are wondering why things go kablooey. Or someone borrows your code, but I guess that's not strictly your problem. Anyways there's a reason people often use somewhat longer variable names.

Although I see your point. I think the worry is unwarranted. All contribution are checked by CI build before it could be merged into master. Even if things could go kablooey, they will be spotted immediately by our CI builds.

**Comment by weitjong** [comment]
> I am starting to think that most targets in a Visual Studio IDE will not in fact build when you select them "Project Only". I've been chasing a lot of things backwards and finding they won't build, and give errors about what they're expecting that doesn't exist yet.

There was recent change in renaming the 'Source/Urho3D/Script' directory to 'Source/Urho3D/AngelScript'. It may be wise to delete and recreate your build tree again just to ensure your build tree is up to date.

Do you have MKLINK privilege setup in your Windows 10 host system? If you don't have it then all the SDK-like include dirs are only copied during the post build step for each of the 3rd party targets. I am not familiar with VS, so I am not sure when you choose "Project Only" for just one target, ScriptCompiler or Urho3D, then will it also try to build those 3rd party targets first or not. I hope it still does, cause if it does not then all those include dirs setup won't work.

**Comment by bvanevery** [comment]
git source directory is on branch Master, up to date, nothing to commit, working directory clean.  Build directory deleted.  CMake started from scratch.  When building only project ScriptCompiler I get:
`1>C:\Users\bvanevery\devel\Urho3D\Source\Tools\ScriptCompiler\ScriptCompiler.cpp(23): fatal error C1083: Cannot open include file: 'Urho3D/AngelScript/Script.h': No such file or directory`
Line 23 is:
`#include <Urho3D/AngelScript/Script.h>` 
Urho3D\Source\Urho3D\AngelScript\Script.h exists just fine, so include path has to be wrong at this point in project configuration.  Or else not designed to work "build only", contrary to my expectation.

Looking in Urho3D\Source\Urho3D\CMakeLists.txt I see some string processing on GLOBAL_INCLUDE_DIRS and ENGINE_INCLUDE_DIRS.

**Comment by cadaver** [comment]
Did a bit more testing. Yes I believe that "project only" build of doc project in a clean checkout cannot be expected to work, as it does not build the required dependency projects and fails on ScriptCompiler. Selecting a normal build of the project (e.g. right-click and select Build) will build the dependencies first and work.

As a sidenote, using the flags URHO3D_64BIT=1 and URHO3D_D3D11=1 I managed to reproduce the HTML help compilation issue. The html files were produced correctly, so it is indeed weird.

Closing this now.

**Comment by weitjong** [comment]
> Did a bit more testing. Yes I believe that "project only" build of doc project in a clean checkout cannot be expected to work, as it does not build the required dependency projects and fails on ScriptCompiler. Selecting a normal build of the project (e.g. right-click and select Build) will build the dependencies first and work.

Yes, I just confirmed that also. "Project only" build does not seem to respect the dependency. It literally just build only the selected project. Interestingly though, invoking the "target only" build via CLI does not have this issue. I use 'rake make target=doc' which in turn calls 'cmake --build' which in turn calls 'MSBUILD' via command line interface to build just 'doc' target. In this manner it seems the dependency of the target is honored.

> As a sidenote, using the flags URHO3D_64BIT=1 and URHO3D_D3D11=1 I managed to reproduce the HTML help compilation issue. The html files were produced correctly, so it is indeed weird.

I could not reproduce the HTML help compilation error issue with this combination also on my Win7 VM. I have thrown in URHO3D_LUAJIT=1 as well into the mix. But as I commented in #910, I see now this issue is caused by faulty external tool. Most probably hhc.exe does not return proper exit status and raise an error handling on Doxygen and in turn in our doc build target.
